### ISSUE TITLE
Using validateTLSCertificates() prevents fetching specific url

### ISSUE THREAD

**Issue by moneytoo** [jhy/jsoup#659]
Fetching ``https://roundcube.net/news/2015/11/23/roundcube-webmail-1.2-beta-out-now/`` the default way works just fine...
```
Document document = Jsoup.connect("https://roundcube.net/news/2015/11/23/roundcube-webmail-1.2-beta-out-now/").get();
```

But in real life, I need to use ``validateTLSCertificates()`` to cover sites using self sign certs, browser specific root CA's. However using ``.validateTLSCertificates(false)`` in 

```
Document document = Jsoup.connect("https://roundcube.net/news/2015/11/23/roundcube-webmail-1.2-beta-out-now/").validateTLSCertificates(false).get();
```
causes
```
Exception in thread "main" javax.net.ssl.SSLException: Received fatal alert: internal_error
	at sun.security.ssl.Alerts.getSSLException(Alerts.java:208)
	at sun.security.ssl.Alerts.getSSLException(Alerts.java:154)
	at sun.security.ssl.SSLSocketImpl.recvAlert(SSLSocketImpl.java:2023)
	at sun.security.ssl.SSLSocketImpl.readRecord(SSLSocketImpl.java:1125)
	at sun.security.ssl.SSLSocketImpl.performInitialHandshake(SSLSocketImpl.java:1375)
	at sun.security.ssl.SSLSocketImpl.startHandshake(SSLSocketImpl.java:1403)
	at sun.security.ssl.SSLSocketImpl.startHandshake(SSLSocketImpl.java:1387)
	at sun.net.www.protocol.https.HttpsClient.afterConnect(HttpsClient.java:559)
	at sun.net.www.protocol.https.AbstractDelegateHttpsURLConnection.connect(AbstractDelegateHttpsURLConnection.java:185)
	at sun.net.www.protocol.https.HttpsURLConnectionImpl.connect(HttpsURLConnectionImpl.java:153)
	at org.jsoup.helper.HttpConnection$Response.execute(HttpConnection.java:512)
	at org.jsoup.helper.HttpConnection$Response.execute(HttpConnection.java:493)
	at org.jsoup.helper.HttpConnection.execute(HttpConnection.java:205)
	at org.jsoup.helper.HttpConnection.get(HttpConnection.java:194)
	at com.company.Main.main(Main.java:166)
```

However using ```.validateTLSCertificates(true)``` doesn't work neither in
```
Document document = Jsoup.connect("https://roundcube.net/news/2015/11/23/roundcube-webmail-1.2-beta-out-now/").validateTLSCertificates(true).get();
```
and causes
```
Exception in thread "main" org.jsoup.HttpStatusException: HTTP error fetching URL. Status=403, URL=https://roundcube.net/news/2015/11/23/roundcube-webmail-1.2-beta-out-now/
	at org.jsoup.helper.HttpConnection$Response.execute(HttpConnection.java:537)
	at org.jsoup.helper.HttpConnection$Response.execute(HttpConnection.java:493)
	at org.jsoup.helper.HttpConnection.execute(HttpConnection.java:205)
	at org.jsoup.helper.HttpConnection.get(HttpConnection.java:194)
	at com.company.Main.main(Main.java:166)
```

I run jsoup 1.8.3 and Java 1.8.0_65.

**Comment by moneytoo** [comment]
It's upstream bug in Java. It was already reported in https://bugs.openjdk.java.net/browse/JDK-8072464 but closed - it's easier to replicate with sites relaying on SNI only. As I can't comment to that bug I filled a new one at Oracle.

**Comment by jhy** [comment]
I can hit it OK in Oracle Java 1.8.0_77 - think this was an issue fixed relatively recently.

**Comment by moneytoo** [comment]
This sample still throws SSLException when HostnameVerifier is used:

```
package com.company;

import javax.net.ssl.HostnameVerifier;
import javax.net.ssl.HttpsURLConnection;
import javax.net.ssl.SSLSession;
import java.io.IOException;
import java.net.URL;

public class Main {
    public static void main(String[] args) throws IOException {
        URL url = new URL("https://roundcube.net/news/2015/11/23/roundcube-webmail-1.2-beta-out-now");
        HttpsURLConnection conn = (HttpsURLConnection) url.openConnection();
        conn.setRequestProperty("User-Agent", "Mozilla/5.0 (Windows NT 6.1; WOW64; rv:40.0) Gecko/20100101 Firefox/40.0");
        conn.setHostnameVerifier(new HostnameVerifier() {
            public boolean verify(String s, SSLSession sslSession) {
                return true;
            }
        });
        conn.getInputStream();
    }
}
```

```
"C:\Program Files\Java\jdk1.8.0_92\bin\java" ...
Exception in thread "main" javax.net.ssl.SSLException: Received fatal alert: internal_error
	at sun.security.ssl.Alerts.getSSLException(Alerts.java:208)
	at sun.security.ssl.Alerts.getSSLException(Alerts.java:154)
	at sun.security.ssl.SSLSocketImpl.recvAlert(SSLSocketImpl.java:2023)
	at sun.security.ssl.SSLSocketImpl.readRecord(SSLSocketImpl.java:1125)
	at sun.security.ssl.SSLSocketImpl.performInitialHandshake(SSLSocketImpl.java:1375)
	at sun.security.ssl.SSLSocketImpl.startHandshake(SSLSocketImpl.java:1403)
	at sun.security.ssl.SSLSocketImpl.startHandshake(SSLSocketImpl.java:1387)
	at sun.net.www.protocol.https.HttpsClient.afterConnect(HttpsClient.java:559)
	at sun.net.www.protocol.https.AbstractDelegateHttpsURLConnection.connect(AbstractDelegateHttpsURLConnection.java:185)
	at sun.net.www.protocol.http.HttpURLConnection.getInputStream0(HttpURLConnection.java:1513)
	at sun.net.www.protocol.http.HttpURLConnection.getInputStream(HttpURLConnection.java:1441)
	at sun.net.www.protocol.https.HttpsURLConnectionImpl.getInputStream(HttpsURLConnectionImpl.java:254)
	at com.company.Main.main(Main.java:20)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at com.intellij.rt.execution.application.AppMain.main(AppMain.java:144)

Process finished with exit code 1
```

The upstream bug JDK-8144566 is still marked ``In Progress`` and it doesn't work in latest stable 1.8.0_92. So unless there's a workaround implemented in jsoup, it's still affected. (I believe I read that it worked in some releases, probably around _77 but regressed again...)

**Comment by moneytoo** [comment]
Using ``socket.getSSLParameters().setEndpointIdentificationAlgorithm("HTTPS")`` should work around that: http://stackoverflow.com/questions/36323704/sni-client-side-mystery#comment60314450_36343704

**Comment by moneytoo** [comment]
I end up with this fix cf1968635c53701d0b6a5b53e5a4282301e0b253
